scylla/src/transport/session.rs

```diff
@@ -148,17 +151,25 @@ impl Session {
     /// # Arguments
     ///#
     pub async fn prepare(&self, query: &str) -> Result<PreparedStatement> {
-        // FIXME: Prepared statement ids are local to a node, so we must make sure
-        // that prepare() sends to all nodes and keeps all ids.
-        let result = self.any_connection()?.prepare(query.to_owned()).await?;
-        match result {
-            Response::Error(err) => Err(err.into()),
-            Response::Result(result::Result::Prepared(p)) => Ok(PreparedStatement::new(
-                p.id,
-                p.prepared_metadata,
-                query.to_owned(),
-            )),
-            _ => Err(anyhow!("Unexpected frame received")),
+        let connections = self.get_connections()?;
+        let all_connections = connections.iter().map(|(_, v)| v).flatten();
+
+        // Prepare statements on all connections concurrently
+        let handles = all_connections.map(|c| c.prepare(query));
+        let mut results = try_join_all(handles).await?;
+
+        // Check if every received statement has the same ID
+        for p in results.windows(2) {
```

Comment:
Please move this additional check to a separate helper function (called validate_prepare_ids or something alike).
Also, consider writing it in a more declarative fashion, e.g.:
```rust
results.windows(2).all(|p| p[0].get_id() == p[1].get_id()) // + add error handling at the end of the chain
```

Reply:
Done